### ISSUE TITLE
Have a pre-commit check

### ISSUE THREAD

**Issue by barbaricyawps** [thegooddocsproject/templates#159]
I recently opened a PR to the incubator and @Loquacity discovered my commit had some "smart apostrophes" in it. One possible workaround for catching this would be to add a pre-commit checker for PRs. However, this is a little tricky to set up and it's tricky to learn how to set it up on your local preview if you're not familiar with that convention.

**Comment by Loquacity** [comment]
Adding some linting to a GH Action would be a great idea, imho :+1:

**Comment by mgan59** [comment]
@barbaricyawps Agreed, think it ties into what @ddbeck is proposing as well in terms of linting for markdown.  Within the incubator/#4 pull-request would help to know which commit has the smart apostrophes so when we build the tooling we can have a test-case.

**Comment by Loquacity** [comment]
> @barbaricyawps Agreed, think it ties into what @ddbeck is proposing as well in terms of linting for markdown. Within the [incubator/#4](https://github.com/thegooddocsproject/incubator/pull/4) pull-request would help to know which commit has the smart apostrophes so when we build the tooling we can have a test-case.

The original commit in the PR had a mix of both, to my eye.

**Comment by ddbeck** [comment]
My source formatter suggestion was in #160, though I don't think Prettier itself enforces a Markdown quotes style. We might need additional process or tooling on top of a source formatter for that sort of stylistic issue.

**Comment by barbaricyawps** [comment]
@ScriptAutomate might be able to weigh in on this issue too, so tagging him here.

**Comment by ScriptAutomate** [comment]
A minimum-recommended pre-commit setup:

### .pre-commit-config.yaml

```yaml
---
default_language_version:
  python: python3

minimum_pre_commit_version: 2.9.2
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v3.3.0
    hooks:
      - id: check-merge-conflict  # Check for files that contain merge conflict strings.
      - id: mixed-line-ending     # Replaces or checks mixed line endings.
        args: [--fix=lf]          # Defaults to *nix-style line endings.
      - id: end-of-file-fixer     # Makes sure files end in a newline and only a newline.
      - id: check-json            # Check json file formatting.
      - id: check-yaml            # Check yaml file formatting.
      - id: detect-private-key    # Ensure no accidental private keys added to repo.
      - id: trailing-whitespace   # Remove unneeded whitespace.
  -   repo: https://github.com/codespell-project/codespell
      rev: v2.0.0
      hooks:
      - id: codespell             # Base spellcheck across repository.
```

### .github/workflows/actions.yaml

```yaml
# This is a basic workflow to help you get started with Actions

name: CI

# Controls when the action will run.
on:
  # Triggers the workflow on push or pull request events but only for the main branch
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

  # Allows you to run this workflow manually from the Actions tab
  workflow_dispatch:

# A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  # This workflow contains a single job called "test"
  test:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v2
      # Runs all the pre-commit tests in .pre-commit-config.yaml
      - name: pre-commit checks
        run: |
          sudo apt-get update && sudo apt-get install python3-venv
          python3 -m venv .venv; source .venv/bin/activate; pip install -U pip setuptools wheel
          pip install pre-commit
          pre-commit install
          pre-commit run -a -v --color always

      # Checks for broken URLs
      - name: Check for broken URL links with brok
        run: |
          sudo apt-get update && sudo apt-get install wget libtinfo5 -y
          wget https://github.com/smallhadroncollider/brok/releases/download/1.1.0/brok-1.1.0_x86-64-linux.deb
          sudo dpkg -i brok-1.1.0_x86-64-linux.deb
          brok --ignore $(cat .brok-ignore-links) $(find . -type f -name "*.md")
```

> Note that I'm only scanning `.md` markdown files with the `brok` example, which can be expanded if we wanted.

### .brok-ignore-links

> These are URLs selected as needing to be skipped in an audit

```
https://example.com
```

That would help with a lot of general annoyances. I'm not sure how to get `codespell` to skip special entries where it is raising false flags, though, so we may not want it if it gives us grief unless someone else can figure out how to work with that.

**We would also need to include directions on how [`pre-commit`](https://pre-commit.com/) works, and how people can set it up locally. The goal of pre-commit is to catch as many problems as possible before you even make a commit to a branch.**

## Smart Quotes / Apostrophes

As for smart quotes/apostrophes replacing: I've put together a little python script that is pre-commit compatible, but it should honestly be sitting in it's own repo. Maybe that can be a separate issue? I think this could really be a way of setting a baseline standard for tests across The Good Docs project, and that we should consider creating a centralized repo for these tests to be housed, unless `.github` may allow us to create GitHub Actions meant to be executed across the board? I imagine many of these would be useful across all TGD projects, especially if we end up automating any style guides or other things we choose to standardize on.
